qiskit/algorithms/evolvers/classical_methods/scipy_real_evolver.py

```diff
@@ -0,0 +1,328 @@
+            lhs_operator,
+            rhs_operator,
+            aux_ops,
+            aux_ops_2,
+            timesteps,
+            self._bicg_tol(timesteps),
+        )
+
+    def _bicg_tol(self, timesteps: int) -> float:
+        """Returns the tolerance for the BiCG solver.
+
+        Args:
+            timesteps: The number of timesteps in the evolution.
+
+        Returns:
+            The tolerance for the BiCG solver.
+        """
+        return self.threshold * self.bicg_err / timesteps
+
+    def minimal_number_steps(
+        self, norm_hamiltonian: float, time: float, threshold: float = 1e-4
+    ) -> int:
+        r"""Calculate the timestep for the given time and threshold.
+
+        We use the fact that the Taylor expansion term of third order in our expansion would be
+        :math:`\frac{ ( -i H \delta t ) ^3}{12}` to compute an approximation for how
+        many timesteps we need to reach a certain precision.
+
+        Args:
+            time: The time to evolve.
+            norm_hamiltonian: The operator norm of the Hamiltonian if known or an estimation of
+                             it (For example the infinity norm of the Hamiltonian). This value
+                             will be associated with the error of the Hamiltonian.
+            threshold: The threshold for the error.
+
+        Returns:
+            The number of timesteps needed to reach the error threshold.
+        """
+        return int(
+            np.power(norm_hamiltonian * time, 3 / 2)
+            * np.power(12 * threshold * (1 - self.bicg_err), -1 / 2)
+            + 1
+        )
+
+    def _ntimesteps(self, time: float, hamiltonian: sp.csr_matrix, threshold: float = 1e-4) -> int:
+        """Calculate the number of timesteps needed to reach the threshold error if the user doesn't
+        indicate the number of timesteps.
+
+        Uses the infinity norm to estimate the operator norm of the Hamiltonian.
+        Returns:
+            The number of timesteps needed to reach the error threshold.
+        """
+        hnorm = norm(hamiltonian, ord=np.inf)
+        return self.minimal_number_steps(norm_hamiltonian=hnorm, time=time, threshold=threshold)
+
+    def _step(
+        self,
+        state: np.ndarray,
+        lhs_operator: sp.csr_matrix,
+        rhs_operator: sp.csr_matrix,
+        bicg_tol: float,
+    ) -> np.ndarray:
+        """ "Perform one timestep of the evolution.
+
+        Args:
+            state: The initial state.
+            timestep: The timestep to evolve.
+            hamiltonian: The Hamiltonian to evolve under.
+
+        Returns:
+            The evolved state.
+
+        Raises:
+            RuntimeError: If the biconjugate gradient solver fails.
+
+        """
+
+        rhs = rhs_operator.dot(state)
+        ev_state, exitcode = bicg(A=lhs_operator, b=rhs, x0=state, atol=bicg_tol, tol=bicg_tol)
+        if exitcode != 0:
+            raise RuntimeError("The biconjugate gradient solver has falied.")
```

Comment:
Should we add the exitcode here so users could try and look up what went wrong?

Reply:
Good point.